cmd/uplink/cmd/cp_test.go

```diff
@@ -4,122 +4,180 @@
+}
+
+var testObjects = createObjects()
+
+// createObjects generates the objects (i.e. slice of bytes) that
+// will be used as the objects to upload/download tests
+func createObjects() map[string][]byte {
+	objects := make(map[string][]byte)
+	for _, bm := range benchmarkCases {
+		data := make([]byte, bm.objectsize)
+		_, err := rand.Read(data)
+		if err != nil {
+			log.Fatalf("failed to read random bytes: %+v\n", err)
+		}
+		objects[bm.name] = data
+	}
+	return objects
+}
 
+// uplinkSetup setups an uplink to use for testing uploads/downloads
+func uplinkSetup() s3client.Client {
+	conf, err := setupConfig()
+	if err != nil {
+		log.Fatalf("failed to setup s3client config: %+v\n", err)
+	}
 	client, err := s3client.NewUplink(conf)
 	if err != nil {
-		log.Fatalf("failed to create NewUplink: %+v\n", err)
+		log.Fatalf("failed to create s3client NewUplink: %+v\n", err)
 	}
 	err = client.MakeBucket(bucket, "")
 	if err != nil {
-		log.Fatalf("failed to create bucket %q: %+v\n", bucket, err)
+		log.Fatalf("failed to create bucket with s3client %q: %+v\n", bucket, err)
 	}
 	return client
 }
 
-func teardown(client s3client.Client) {
-	for _, bm := range benchmarkCases {
-		deleteUploadedFiles(bm.name)
-	}
-
-	err := client.RemoveBucket(bucket)
-	if err != nil {
-		log.Fatalf("failed to remove bucket %q: %+v\n", bucket, err)
+func setupConfig() (s3client.Config, error) {
+	const (
+		uplinkEncryptionKey  = "supersecretkey"
+		defaultSatelliteAddr = "127.0.0.1:10000"
+	)
+	var conf s3client.Config
+	conf.EncryptionKey = uplinkEncryptionKey
+	conf.Satellite = getEnvOrDefault("SATELLITE_0_ADDR", defaultSatelliteAddr)
+	conf.APIKey = getEnvOrDefault(os.Getenv("GATEWAY_0_API_KEY"), os.Getenv("apiKey"))
+	if conf.APIKey == "" {
+		return conf, errors.New("no api key provided. Expecting an env var $GATEWAY_0_API_KEY or $apiKey")
 	}
+	return conf, nil
 }
 
-func deleteUploadedFiles(name string) {
-	for filePath := range uploadedFiles[name] {
-		err := client.Delete(bucket, filePath)
-		if err != nil {
-			log.Printf("failed to delete file %q: %+v\n", filePath, err)
-		}
+func getEnvOrDefault(key, fallback string) string {
+	if value, exists := os.LookupEnv(key); exists {
+		return value
 	}
+	return fallback
 }
 
-var client = setup()
-
-var benchmarkCases = []struct {
-	name     string
-	filesize memory.Size
-}{
-	{"100B", 100 * memory.B},
-	{"1MB", 1 * memory.MiB},
-	{"10MB", 10 * memory.MiB},
-	{"100MB", 100 * memory.MiB},
-	{"1G", 1 * memory.GiB},
-}
+func BenchmarkUpload(b *testing.B) {
+	var client = uplinkSetup()
 
-var uploadedFiles = map[string]map[string]struct{}{}
+	// uploadedObjects is used to store the names of all objects that are uploaded
+	// so that we can make sure to delete them all during cleanup
+	var uploadedObjects = map[string][]string{}
 
-func BenchmarkUpload(b *testing.B) {
 	for _, bm := range benchmarkCases {
 		b.Run(bm.name, func(b *testing.B) {
-			file := createRandomBytes(bm.filesize)
 			b.SetBytes(1)
```

Comment:
I'm curious about this call. Would it be possible to do something like `b.SetBytes(len(testObjects[bm.name]))`, so that the reported MB/s is accurate? Or if we don't care about the MB/s or ns/op measurements, why call `SetBytes()` at all? I admit there's a lot about the testing lib I don't know.

Reply:
oops! great catch and yes that is embarrassing of course I need to set the actual number of bytes hahahaha. _fixing_